week 2 assignment 1


submitted by: Shruti Gupta


branch: IT2


enrollment number:12301032025



In [1]:
#importing libraries
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LinearRegression

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.preprocessing import LabelEncoder


Part A: Dataset Understanding

In [2]:
#1. Load the dataset and display the first five records.
df=pd.read_csv("/content/Dataset 2.csv")
print(df.head())

   UserID  Age  Gender SubscriptionType  WatchHoursPerWeek  DevicesUsed  \
0    1001   22  Female            Basic                 23            5   
1    1002   55    Male            Basic                  9            4   
2    1003   49    Male            Basic                  8            3   
3    1004   39  Female          Premium                 19            5   
4    1005   38  Female          Premium                 23            5   

  FavoriteGenre  AdClicks  MonthlySpend SubscriptionRenewed  
0        Comedy        13           353                  No  
1         Drama        14           317                 Yes  
2        Comedy        16           309                  No  
3         Drama        45           833                 Yes  
4        Sci-Fi        24           804                 Yes  


In [3]:
#2. Determine the number of rows and columns in the dataset.
df.shape

(750, 10)

In [4]:
#3. Display all column names.
df.columns


Index(['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek',
       'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend',
       'SubscriptionRenewed'],
      dtype='object')

In [6]:
#4. Identify numerical and categorical features.
df.dtypes

,0
UserID,int64
Age,int64
Gender,object
SubscriptionType,object
WatchHoursPerWeek,int64
DevicesUsed,int64
FavoriteGenre,object
AdClicks,int64
MonthlySpend,int64
SubscriptionRenewed,object


In [8]:
#5. Check whether the dataset contains missing values.
print("\nMissing values: ")
print(df.isnull().sum())
print("\nTotal Missing Values:",df.isnull().sum().sum())


Missing values: 
UserID                 0
Age                    0
Gender                 0
SubscriptionType       0
WatchHoursPerWeek      0
DevicesUsed            0
FavoriteGenre          0
AdClicks               0
MonthlySpend           0
SubscriptionRenewed    0
dtype: int64

Total Missing Values: 0


Part B: Exploratory Data Analysis


In [10]:
#6.Calculate the average age of users.
float(df["Age"].mean())

41.824

In [11]:
#7.Determine the average watch hours per week.
print(df['WatchHoursPerWeek'].mean())

14.236


In [12]:
#8. Find the average monthly spending of users.
df['MonthlySpend'].mean()

np.float64(689.9053333333334)

In [13]:
#9.Count the number of users in each subscription category.
df['SubscriptionType'].value_counts()

,count
SubscriptionType,
Basic,342
Premium,279
VIP,129


In [14]:
#10. Determine the percentage of users who renewed their subscriptions.
(df['SubscriptionRenewed'].value_counts(normalize=True)*100)

,proportion
SubscriptionRenewed,
No,53.733333
Yes,46.266667


Part C: Data Preparation


In [16]:
#11.Convert categorical features into numerical form.

df_encoded=df.copy()
le=LabelEncoder()
for col in df_encoded.select_dtypes(include='object').columns:
    df_encoded[col]=le.fit_transform(df_encoded[col])
df_encoded.head()

,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,0,0,23,5,1,13,353,0
1,1002,55,1,0,9,4,2,14,317,1
2,1003,49,1,0,8,3,1,16,309,0
3,1004,39,0,1,19,5,2,45,833,1
4,1005,38,0,1,23,5,5,24,804,1


In [19]:
#12. Define the feature set (X) and target variable (y) for subscription renewal prediction.
X=df_encoded.drop(['UserID','SubscriptionRenewed'],axis=1)
y=df_encoded['SubscriptionRenewed']


In [20]:
#13. Split the dataset into training and testing sets.
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

Part D: Decision Tree Classification


In [21]:
#14.Train a Decision Tree model to predict whether a user will renew their subscription.
dt_model = DecisionTreeClassifier(random_state=42)
dt_model.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

In [23]:
#15.Evaluate the model using accuracy.
y_pred_dt = dt_model.predict(X_test)
dt_accuracy = accuracy_score(y_test,y_pred_dt)
print("Decision Tree Accuracy:", dt_accuracy)

Decision Tree Accuracy: 0.56


In [24]:
#16.Generate and interpret the confusion matrix.
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred_dt)
print(cm)

[[48 34]
 [32 36]]


Part E: K-Nearest Neighbors (KNN)


In [25]:
#17.Train a KNN classifier with K = 5.
knn_model= KNeighborsClassifier(n_neighbors=5)
knn_model.fit(X_train, y_train)
y_pred_knn= knn_model.predict(X_test)

In [26]:
#18.Compare the accuracy of KNN with the Decision Tree model.
knn_accuracy=accuracy_score(y_test, y_pred_knn)
print("Decision Tree Accuracy:",dt_accuracy)
print("KNN Accuracy:",knn_accuracy)

Decision Tree Accuracy: 0.56
KNN Accuracy: 0.6333333333333333


Part F: Linear Regression


In [27]:
#19.Train a Linear Regression model to predict monthly spending.
X_reg=df_encoded.drop(['UserID','MonthlySpend'], axis=1)
y_reg=df_encoded['MonthlySpend']

X_train_reg, X_test_reg, y_train_reg, y_test_reg=train_test_split(
    X_reg, y_reg, test_size=0.2, random_state=42
)
lr_model=LinearRegression()
lr_model.fit(X_train_reg, y_train_reg)

LinearRegression()

In [28]:
#20.Predict the monthly spending for a new user and interpret the result.
new_user=[30,1,2,15,3,0,10,1]
prediction= lr_model.predict([new_user])
print("Predicted Monthly Spend:",prediction[0])

Predicted Monthly Spend: 1318.3732197519025


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


1. Which factors appear to influence subscription renewal the most?

answer: subscription type, watch hours per week, monthly spending, and advertisement clicks

2. Why is subscription renewal a classification problem?

answer :Because the target variable has two categories: Yes or No.

3. Why is monthly spending a regression problem?

answer: Because monthly spending is a continuous numerical value.

4. Which algorithm performed better for renewal prediction?

answer: The algorithm with the higher accuracy score  performed better i.e. KNN Accuracy with 0.63333333 of accuracy score.


5. How could the platform use these predictions to improve customer retention?

answer: Netflix can identify users likely to cancel and offer personalized recommendations, discounts, or subscription upgrades to retain them.
